libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt

```diff
@@ -0,0 +1,296 @@
+                LocationBufferCommands.BUFFER_READ -> {
+                    locationBuffer.forEach { location ->
+                        locationAction.predicate(location)
+                    }
+                }
+            }
+        }
+
+        /**
+         * This monitors the life time of the job that handles the channels. If that job is terminated, we cleanup.
+         */
+        jobControl.scope.launch {
+            select<Unit> {
+                cleanupJob.onJoin {
+                    locationEngine.removeLocationUpdates(locationCallback)
+                }
+            }
+        }
+    }
+
+    private suspend fun getDatePartitionedLocations(predicate: (Location) -> Boolean): Array<Location> {
+        val retVal = mutableListOf<Location>()
+        channelLocationBuffer.send(LocationBufferControl(LocationBufferCommands.BUFFER_READ, null) { location ->
+            if (predicate(location)) {
+                retVal.add(location)
+            }
+        })
+        return retVal.toTypedArray()
+    }
+
+    private fun parseRouteProgressAsync(): Deferred<TelemetryStep> {
+        return jobControl.scope.async {
+            val telemetryStep = TelemetryStep() // Default initialization
+            val routeProgress = channelOnRouteProgress.receive()
+            routeProgress.currentLegProgress()?.upcomingStep()?.let { legStep ->
+                telemetryStep.distance = legStep.distance().toInt()
+                telemetryStep.distanceRemaining = routeProgress.distanceRemaining().toInt()
+                telemetryStep.duration = legStep.duration().toInt()
+                telemetryStep.durationRemaining = routeProgress.durationRemaining().toInt()
+                telemetryStep.previousModifier = routeProgress.currentLegProgress()?.currentStepProgress()?.step()?.maneuver()?.modifier()
+                        ?: ""
+                telemetryStep.previousName = routeProgress.currentLegProgress()?.currentStepProgress()?.step()?.name()
+                        ?: ""
+                telemetryStep.previousType = routeProgress.currentLegProgress()?.currentStepProgress()?.step()?.maneuver()?.type()
+                        ?: ""
+                telemetryStep.upcomingType = legStep.maneuver().type() ?: ""
+                telemetryStep.upcomingModifier = legStep.maneuver().modifier() ?: ""
+                telemetryStep.upcomingName = legStep.name() ?: ""
+            }
+            telemetryStep
+        }
+    }
+
+    /**
+     * TODO:OZ add one of these for each telemetry event
+     * Generates a well-formed Feedback event.
+     */
+    private suspend fun populateUserFeedbackEvent(event: TelemetryEventFeedback): TelemetryEventInterface {
+        val phoneState = PhoneState(context)
+
+        val feedbackEvent = TelemetryUserFeedbackWrapper(event,
+                phoneState.userId,
+                phoneState.audioType,
+                getDatePartitionedLocations { location -> location.time < System.currentTimeMillis() },
```

Comment:
We currently (legacy) track 20 locations before and after the event and now we're only storing 20 👀 https://github.com/mapbox/mapbox-navigation-android/blob/662a6fc73ecc9d0f89bb5ef9f2e2f367cf3d99eb/libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt#L49

Reply:
👍 This code has been removed. There are now two queues, each holds a max of 20 locations